libs/x-content/src/main/java/org/elasticsearch/xcontent/FlatteningXContentParser.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.xcontent;
+
+import java.io.IOException;
+
+/**
+ * A subclass of XContentSubParser that provides the functionality to flatten
+ * the field names by prefixing them with the provided parent name.
+ */
+public class FlatteningXContentParser extends XContentSubParser {
```

Comment:
Sorry for chiming in late, I'd like to better understand how the additional complexity in `ContentPath` around tracking the full path is replaced with this custom parser. In a previous experience with `DotExpandingXContentParser`, we have seen that having a custom parser can have quite a few side effects that are difficult to predict (do we forward all the right methods? is the parser only used how we think it is?),  because the parser exposes a pretty generic interface that is used in many different ways and we are customizing it for a very specific scenario.

While additional methods to `ContentPath` increase its surface, I find that they make the contract easier to reason about and easier to track as well. While I agree that `ContentPath` is not a fantastic abstraction and needs some love, I am not sure that adding a new parser addresses that fairly, in that it introduces all kinds of unknowns, at least in my opinion.

Reply:
> because the parser exposes a pretty generic interface that is used in many different ways and we are customizing it for a very specific scenario.

That's a fair point - maybe we should make this a private class within DocumentParserContext?

My reason for preferring it this way is that is makes it much simpler to read the DocumentParser code, which is already fairly gnarly.  It also opens the way to further simplifications - I played around a bit with rewriting the Dynamic.RUNTIME logic to use this and it mostly just works, although we'd need to revisit how we handle fields that contain spaces.

Reply:
I agree that DocumentParser could be simplified, greatly. But I think that this type of changes just end up moving the complexity elsewhere and possibly cause collateral issues that are difficult to predict. I am being conservative, I know, that's because I've been bitten with DotExpandingXContentParser before.